smartsim/settings/dispatch.py

```diff
@@ -0,0 +1,365 @@
+        :raises errors.LauncherNotFoundError: No compatible launcher was
+            yielded from the provided iterator.
+        :returns: A launcher configured with the provided settings.
+        """
+        launcher = helpers.first(self._is_compatible_launcher, from_available_launchers)
+        if launcher is None:
+            raise errors.LauncherNotFoundError(
+                f"No launcher of exactly type `{self.launcher_type.__name__}` "
+                "could be found from provided launchers"
+            )
+        return self.create_adapter_from_launcher(launcher, with_settings)
+
+
+@t.final
+class _LauncherAdapter(t.Generic[Unpack[_Ts]]):
+    """An adapter class that will wrap a launcher and allow for a unique
+    signature of for its `start` method
+    """
+
+    def __init__(
+        self,
+        launcher: LauncherProtocol[_LaunchableT],
+        map_: t.Callable[[Unpack[_Ts]], _LaunchableT],
+    ) -> None:
+        """Initialize a launcher adapter
+
+        :param launcher: The launcher instance this class should wrap
+        :param map_: A callable with arguments for the new `start` method that
+            can translate them into the expected launching type for the wrapped
+            launcher.
+        """
+        # NOTE: We need to cast off the `_LaunchableT` -> `Any` in the
+        #       `__init__` method signature to hide the transform from users of
+        #       this class. If possible, this type should not be exposed to
+        #       users of this class!
+        self._adapt: t.Callable[[Unpack[_Ts]], t.Any] = map_
+        self._adapted_launcher: LauncherProtocol[t.Any] = launcher
+
+    def start(self, *args: Unpack[_Ts]) -> LaunchedJobID:
+        """Start a new job through the wrapped launcher using the custom
+        `start` signature
+
+        :param args: The custom start arguments
+        :returns: The launched job id provided by the wrapped launcher
+        """
+        payload = self._adapt(*args)
+        return self._adapted_launcher.start(payload)
+
+
+DEFAULT_DISPATCHER: t.Final = Dispatcher()
+# Disabling because we want this to look and feel like a top level function,
+# but don't want to have a second copy of the nasty overloads
+# pylint: disable-next=invalid-name
+dispatch: t.Final = DEFAULT_DISPATCHER.dispatch
+
+
+# >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
+# TODO: move these to a common module under `smartsim._core.launcher`
+# -----------------------------------------------------------------------------
+
+
+def create_job_id() -> LaunchedJobID:
+    return LaunchedJobID(str(uuid.uuid4()))
+
+
+class ExecutableProtocol(t.Protocol):
+    def as_program_arguments(self) -> t.Sequence[str]: ...
+
+
+class LauncherProtocol(t.Protocol[_T_contra]):
+    def start(self, launchable: _T_contra, /) -> LaunchedJobID: ...
+    @classmethod
+    def create(cls, exp: Experiment, /) -> Self: ...
+
+
+def make_shell_format_fn(
```

Comment:
possible short docstring here

Reply:
Ahh! this seems like an even better spot to add [this docstring](https://github.com/CrayLabs/SmartSim/pull/620#discussion_r1693470855)! Will add!!

Reply:
Docstring w/ example usage has been added to `make_shell_format_fn`

Reply:
Docstring w/ example usage has been added to `make_shell_format_fn`!!